### ISSUE TITLE
Feature: allow lyrics editing in the Lyrics tab

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#32]
Can store them inside the MP3 file (ID3 unsynchronized lyrics)? Or, in the Sqlite DB?

**Comment by mpogue2** [comment]
It's really super annoying to have to edit Lyrics as .TXT or by hand as .HTM ...adding the "SUPER ANNOYING" label to this one.  I have code that did this, but it should be re-written, I think.

**Comment by mpogue2** [comment]
I was prepping for a hoedown for the last couple of days.  The lack of a built-in Lyrics editor was a major pain point for me. 
 
I am currently using Kompozer (free on Mac OS X), and cut/pasting from other lyrics that I already had prepared.  It still got it wrong sometimes, and I had to hand-edit the HTML source.  After I saved a change, I had to go back to SquareDesk, choose another song, and choose the original song again, to get the lyrics reloaded.  PAINFUL.  This process worked, but it was time-consuming and error-prone.

I really wanted to be able to:
1) open an existing HTML cuesheet
2) edit the title section
3) edit the body freely (mostly cut/copy/paste)
4) highlight and set/unset the lyrics spans (<span class="lyrics"></span>), maybe with just a simple button (it should work exactly like a rich text editor's existing Bold, Italic, Underline buttons)
5) same as item 4 above, except for setting the HEADER (used for "OPENER", etc.)
6) save (don't overwrite the existing cue sheet)
7) Have the new cuesheet appear in the Lyrics dropdown immediately, as the new default

I think this would have saved me about an hour of prep time overall.

**Comment by danlyke** [comment]
Okay, this seems pretty easy. A few questions:

1. Do we give full user control over saving, or just create a "UserCuesheets" directory and enforce the name there, probably with a save button. That save overwrites whatever was there.
2. I figure for automatic section finding we can probably just look for "opener|break|closer|tag|figure" in headings. Basically allow two markup levels: header and body, and maybe bold for calls in the body. (We could also have separate buttons for each of these markup types)
3. Simple toggle buttons for formatting, just like any word processor.

**Comment by danlyke** [comment]
Just started hooking up signals and throwing a few controls up in the `editablelyrics` branch. Bold and Italic work.

**Comment by mpogue2** [comment]
I love it already!  Suggestions: italics button should be like the Bold button (small roundish, italic I on it).

My answers to your questions:

1. I think we've should establish a "lyrics" folder (I already use one for the massive cue sheet downloads).  I also find I am sometimes using cue sheets in the same directory as the singing call.  
Perhaps it's like this:  "minimal file system awareness" -- when it's selected by the dropdown (which does show paths), it can be edited.  Actually editing it needs to put a "*" somewhere, if the lyrics are edited at all.  Or, maybe the SAVE button just becomes active.  Then, Save (or Floppy Disk icon button?) would save the changes (perhaps making an auto-backup somewhere?  That would allow one level of Revert Changes, which should be a menu item, but not a button, I think).

We could do an "unlock to edit button", but I think more people are familiar with the "*" when edited, then use SAVE to save it, or SAVE AS to give it a new name.  Default directory for Save As should be the original cue sheet's directory.  When saved, it should add itself to the dropdown, if it matches the usual RegEx.  This makes it consistent with the Playlist change I just made, too.  (Saved Playlist is a Playlist that shows up in the MRU list.  The Lyrics Dropdown is our MRU equivalent for Lyrics cue sheets that match.)

2) The more I think about it, the more I think that Automatic section finding should be done just by regex.  Any line of text that contains (case-insensitive) OPENER is an opener line.  User can select HEADER style for them, if they want, by highlighting and clicking HEADER.  Auto-scrolling should pay attention to the text (e.g. Opener), but it should look at text, not at formatting, IMHO.  So, I don't think we need any special editor features for Auto-sections yet.

Later on, we might consider a "fix it up nice" button, that auto-detects Opener, Middle Break, Closer, Figure <N> and makes them HDR, auto detects title, makes it title, etc.  But I think this would be a one-shot deal for something coming in as .txt, converting to .htm, "fix it up", do more editing, and then Save As .htm .  Helps with the Import process, in fact it could be a Lyrics > Import, or Lyrics > Convert from .txt to .htm menu item.  (Maybe later...)

I think format should follow SquareView exactly, so that resulting cue sheets can be pushed back to SqView (we're all in this together!).  If users want Bold and Italic I'm also OK with that, but I don't think we need a CALLS style, because it's not the SquareView standard.  Calls are just <body>.  SqView only has: {lyrics, hdr, title, artist, label}.  Everything else (including calls) is {body}.  I believe that all SqView spans are done with  <FONT CLASS="lyrics"></FONT>, no actual spans or divs.  This is semantic tagging, and all presentation styling (color, size, etc.) is done just with the CSS file (yellow background, bold title, blue artist & label, red header, black on pink for lyrics), so let's keep that standard, I think.

Furthermore, and this may be controversial, if a user wants to muck around with text sizes, or tables, or or...I think that stuff is best done with an external html editor.

3. Yes!  I think buttons would be exactly the semantic classes: {lyrics, hdr, title, artist, label}, plus Bold and Italic.  I don't THINK we need a "clear format" button, because if the buttons work like MS Word, then highlighting a mixed bold and regular section would highlight the BOLD button, and pressing once would give all BOLD, and pressing again would give NONE BOLD.  I think that's probably the way everybody does it.  So, no special button to clear formats is needed.

Ok, that's my 2 cents...

**Comment by mpogue2** [comment]
Hmmm...testing out that stuff from item 3 above doesn't currently work.
Make one word bold and the next word not.  Selecting both will remove the bold formatting right now...

**Comment by danlyke** [comment]
Okay: 674011e6f6d107170684432a7dc4c70542bcc93f

1. This defaults to the current music path, and then remembers the location you saved stuff. With the base name of the current song file. This is wrong, but at least it's something to fix (though the fix might just be to save the default music path)

2. Yes. I really like this. We're going to have to be smart about "closer" and what sort of headings we search for it in. I need to do some poking around about how to get text in and out of the control. And the "heading" button isn't working because I don't yet know how to do semantic markup and my attempt at just changing font size crashed.

3. The buttons *should* react to whatever type of text the caret is in. This makes marked stuff that crosses boundaries a little weird, but I think the behavior I just uploaded works.

**Comment by danlyke** [comment]
Oh, and:

2. We should probably make some hashtaggable override, ie: use "#closer" to force the closer section, just in case something in the lyrics would otherwise be caught by a regex. But I'm a huge fan of "semantic markup confuses the users".

and it'd be nice to designate a section where we can change out figures, once the figure management gets better.

**Comment by danlyke** [comment]
On your auto-format thing: We should be able to copy and paste from a ceder.net cuesheet and get gold. That's the standard.

**Comment by danlyke** [comment]
Today's productivity brought to you by long waits on paying work processes...

Header button now takes the current selection and wraps it in <h1></h1> tags. Not the right thing, but again, something to fix.

Last save path is saved and loaded.

I realize that we're going to have to deal with issues of when something has been saved and when it hasn't. It's way too easy to quit without saving, but we don't just want to save over the original cuesheets.

**Comment by danlyke** [comment]
And all this goodness is in fd84cd7da0cdb68290d4e11fee1220acc881d22e in the editablelyrics branch. And I'm off to Palm Springs for the IAGSDC convention tomorrow.

**Comment by mpogue2** [comment]
Is there a way for us to increase the number of these: "Today's productivity brought to you by long waits on paying work processes..."??? :-)  Awesomely productive!

**Comment by mpogue2** [comment]
OK, here's what I found so far:
- B and I buttons work as I expected...nice!
- Header changes to bold and big, but not to the right color (red)
- Initial Save Directory doesn't seem to work.  It start in the App directory itself on Mac OS X (which is hidden).
- After a Save, the new filename DOES show up in the DropDownList.  But, the DropDownList also needs to change to that item as its currently-selected item.  Otherwise, I am seeing item 1 in the DropDown, and item 2 (just saved) in the QTextEdit pane.
- The I button is still styled differently from the B button.

This is already feeling like it will save me a LOT of time.

**Comment by mpogue2** [comment]
I have some ideas for the Quit-without-saving question:
a) if filename was "MP123 - foo bar.html", on Quit should save the current version as "MP123 - foo bar.edited.html", in the same directory.  If that file already exists, use "MP123 - foo bar.edited.2.html".  That way, the work isn't lost.  It will show up in the DropDown list, and so editing can continue, if desired.  When done editing, the user can just save it under the original "MP123 - foo bar.html" name (in which case the edited version goes away), or under some other "final" name, in which case the ".edited<N>.html" version sticks around (and can be deleted manually via Finder, if needed).
b) It would be nice to have at least one level of UNDO to get rid of a single change, and  "Lyrics > Revert changes" to get rid of all changes (in this editing session).

**Comment by danlyke** [comment]
Don't know what to tell you about the B and I buttons, they're
identical here. Maybe work your magic in designer?

Header button is just wrapping things in an <h1>. This is wrong.
Probably some CSS stuff would be good, I'm still trying to figure out
how to attach meta information in the editor, and how to read it.

Initial save directory... I did make a final change to try to save and
restore it. Once you save once in the app, is it staying that way while
the app is open?

We should be re-loading the cue sheet after we save it. I beat this
one pretty hard, don't know what's going on, will dig, I guess.
 
Dan

**Comment by danlyke** [comment]
b) control-z should work (does on my platform), and Lyrics->Revert
Changes would be a matter of re-loading the current cuesheet (ie:
trivial).

**Comment by mpogue2** [comment]
I just sent you email...
I have fixes for the I button, it's a workaround for a Mac OS X bug.
I have a quick fix for the header -- I looked at what SqView is doing, and I'd like to match...
Should I commit these small changes, or should I just send to you?

**Comment by mpogue2** [comment]
Or I could also branch, if you want to look them over before integrating...

**Comment by mpogue2** [comment]
Control Z needs to be pressed twice to revert a change to Header span.  That was my problem....so I guess it does work!

**Comment by mpogue2** [comment]
Initial Save Directory: no, once I Save, it does not remember that new location the next time.

**Comment by mpogue2** [comment]
OK, I pushed 770046be681968de4ae165cbaa4fc3132de66aec , which fixes the Italics button on Mac OS X, makes some esthetic changes to the button sizes, and also changes from <H1> to <p CLASS="hdr"> as per SqView convention.  If you prefer something else, we can always revert...

**Comment by mpogue2** [comment]
Copy/paste do paste in the copied text, but it's the wrong size on Mac OS X (too small)....
Oh, interesting.  When I save it, the font size fixes itself.  How strange!

**Comment by mpogue2** [comment]
On my machine, the lastCuesheetSavePath is always "." -- I'm not sure how it got that way.
I also think it should be saved to Settings as soon as the Save is done, rather than when the window closes.  There is similar code in on_actionSave_Playlist_triggered() that does this path remembering, so feel free to copy that, just change the key...

**Comment by mpogue2** [comment]
And, here's the info on the SqView convention for spans:
Header: &lt;p CLASS="hdr"&gt; ... &lt;/p&gt;
Lyrics: &lt;font CLASS="lyrics"&gt; ... &lt;/font&gt;
Title: &lt;font CLASS="title"&gt; ... &lt;/font&gt;
Label: &lt;font CLASS="label"&gt; ... &lt;/font&gt;
Artist: &lt;font CLASS="artist"&gt; ... &lt;/font&gt;

**Comment by mpogue2** [comment]
re: the DropDown not matching the QTextEdit contents after saving -- I think it IS reloading the contents.  It's the DropDown that has not been updated to match.

Also, a feature request:  I am now thinking that the DropDown would be nice if it were sorted.  I saved my test edits as "RIV123 - Summer Sounds 1.html", "RIV123 - Summer Sounds 2.html",etc and in the dropdown list they all show up in random order.

**Comment by mpogue2** [comment]
Another thought:  maybe there should be a "Save" button as well as a "Save As..." button?  Both buttons would be active only if changes were actually made.  There is lots of room in the new editing toolbar...

And menu items to match:  Lyrics > Save, Lyrics > Save As..., Lyrics > Revert Changes.  I'm not sure where the Bold, Italic, Header, Lyrics should go as menu items.  Edit menu, or Lyrics menu?

**Comment by mpogue2** [comment]
Even with it just starting to work, and with minimal features, this is SO useful already.  Oh, how much time I wasted on Komposer on Friday and Saturday!  And, do you know what Jim O is telling people do (it involves manually editing and inserting spaces by hand, using a Windows-only program)?  

It's horrifying to me how much time callers have been spending on getting the lyrics edited so they can be usable, using multiple programs.  I'm looking forward to this feature!

**Comment by mpogue2** [comment]
One more feature request:  when in the Compact View, I think the toolbar for editing should also disappear.  That way, the look is very clean for use at a festival or hoedown, and the size of the pane is maximized.

**Comment by danlyke** [comment]
As mentioned OOB, please patch as you see fit!

**Comment by danlyke** [comment]
Dropdown list should be sorted in relevant priority order, but... we
may want to relook at that now that we're doing fancier stuff.

**Comment by danlyke** [comment]
Okay, compact view hides the editing toolbar now.

**Comment by danlyke** [comment]
And fixed the cuesheet load after save.

**Comment by mpogue2** [comment]
Load after Save is fixed, though!

**Comment by mpogue2** [comment]
Hmmm....there's a more fundamental problem with the current QTextEdit method, I think.  I was playing around with the lyrics editor, and if you look directly at the Save... output, it is a huge inline CSS expansion of the input HTML + stylesheet that was used.  But, I think what we need is the original HTML separate from the stylesheet that was used, so that the user can still change the stylesheet and get a different result for all files.  The way it is right now, once an HTML file is edited, the style sheet that we used internally is essentially applied to that HMTL file forever....

Example original:
```
<body>
<p class="hdr">(16 beat intro)</p>
<p class="hdr">OPENER (Basic)</p>

<span style="font-weight: bold;">Circle left!</span><br>

      <span class="lyrics">Listen to the music of the carousel<br>
        The ting-a-ling-a-ling of the ice cream bell</span><br>

      Do an Allemande Left your Corner, Turn a Right hand 'round your own<br>


      And the Men Star Left and you roll it, roll it around you run<br>

      Turn the Partner Right, then you go Left Allemande<br>

      Swing a little girl and you Promenade the land<br>

      <span class="lyrics">Here...come...summer sounds, the summer sounds I love</span><font class="lyrics"><br>
</font><br>
```

Example Saved Output from QTextEdit:
```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta name="qrichtext" content="1" /><title>Summer Sounds</title><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body style=" font-family:'.SF NS Text'; font-size:13pt; font-weight:400; font-style:normal;" bgcolor="#ffffe0">
<p style=" margin-top:12px; margin-bottom:12px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; line-height:100%; background-color:#ffffe0;"><span style=" font-family:'Verdana'; font-size:x-large; color:#ff0000; background-color:#ffffe0;">(16 beat intro) V 10</span></p>
<p style=" margin-top:12px; margin-bottom:12px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; line-height:100%; background-color:#ffffe0;"><span style=" font-family:'Verdana'; font-size:x-large; color:#ff0000; background-color:#ffffe0;">OPENER (Basic)</span><span style=" font-family:'Verdana'; font-size:large; color:#000000;"> </span><span style=" font-family:'Verdana'; font-size:large; font-weight:600; color:#000000;">Circle left!</span><span style=" font-family:'Verdana'; font-size:large; color:#000000;"><br /></span><span style=" font-family:'Verdana'; font-size:large; color:#000000; background-color:#ffc0cb;">Listen to the music of the carousel<br />The ting-a-ling-a-ling of the ice cream bell</span><span style=" font-family:'Verdana'; font-size:large; color:#000000;"><br />Do an Allemande Left your Corner, Turn a Right hand 'round your own<br />And the Men Star Left and you roll it, roll it around you run<br />Turn the Partner Right, then you go Left Allemande<br />Swing a little girl and you Promenade the land<br /></span><span style=" font-family:'Verdana'; font-size:large; color:#000000; background-color:#ffc0cb;">Here...come...summer sounds, the summer sounds I love<br /></span><span style=" font-family:'Verdana'; font-size:large; color:#000000;"><br /></span></p>
```
I tried running HTML Tidy on it, and tidy was able to factor out all of the spans to a single CSS block, but then we'd need to pattern match the extracted CSS block against the original style sheet that we used.  That's not simple.  Example output from tidy (it's much closer to the original):
```
<STYLE type="text/css">
  body {
  background-color: #FFFFE0;
  }
  body.c7 {font-family:'.SF NS Text'; font-size:13pt; font-weight:400; font-style:normal;}
  p.c6 {margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;}
  span.c5 {font-family:'Verdana'; font-size:large; color:#000000; background-color:#ffc0cb;}
  span.c4 {font-family:'Verdana'; font-size:large; color:#000000;}
  span.c3 {font-family:'Verdana'; font-size:large; font-weight:600; color:#000000;}
  p.c2 {margin-top:12px; margin-bottom:12px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; line-height:100%; background-color:#ffffe0;}
  span.c1 {font-family:'Verdana'; font-size:x-large; color:#ff0000; background-color:#ffffe0;}
  </STYLE>
....
<BODY class="c7">
  <P class="c2"><SPAN class="c1">(16 beat intro) V 10</SPAN></P>
  <P class="c2"><SPAN class="c1">OPENER (Basic)</SPAN> <SPAN class="c3">Circle left!</SPAN><SPAN class="c4"><BR></SPAN><SPAN class="c5">Listen to the music of
  the carousel<BR>
  The ting-a-ling-a-ling of the ice cream bell</SPAN><SPAN class="c4"><BR>
  Do an Allemande Left your Corner, Turn a Right hand 'round your own<BR>
  And the Men Star Left and you roll it, roll it around you run<BR>
  Turn the Partner Right, then you go Left Allemande<BR>
  Swing a little girl and you Promenade the land<BR></SPAN><SPAN class="c5">Here...come...summer sounds, the summer sounds I love<BR></SPAN><SPAN class=
  "c4"><BR></SPAN></P>
  <P class="c2"><SPAN class="c1">FIGURE 1 (Basic, corner progression)</SPAN></P>
  <P class="c6"><SPAN class="c4">(HEADS) Star Thru, Pass Thru<BR>
  Right And Left Thru<BR>
  (you) Veer Left and Bend the Line and make a line you do<BR>
  Pass Thru Wheel and Deal<BR>
  Centers Star Thru<BR>
  Pass Thru Cloverleaf<BR>
  (New Center Two) Square Thru 3/4<BR>
  Turn Corner By the Left<BR>
  All The Way Around and PROMENADE THE SET<BR>
  <BR></SPAN><SPAN class="c5">Here...come...summer sounds, the summer sounds I love<BR></SPAN></P>
```

Compare that to the original cuesheet.css file:
```
body, p, font { font-family: Verdana; font-size: large; font-weight: Normal; line-height: 100%; background: #FFFFE0; color: #000000; }
.lyrics   { font-family: Verdana; font-size: large; font-weight: Normal; color: #000000; background-color: #FFC0CB;}
.hdr      { font-family: Verdana; font-size: x-large; font-weight: Normal; color: #FF0000;}
.title    { font-family: Verdana; font-size: x-large; font-weight: Bold; color: #000000;}

.artist    { font-family: Verdana; font-size: medium; font-weight: Normal; color: #0000FF;}
.label     { font-family: Verdana; font-size: medium; font-weight: Normal; color: #0000FF;}
```

So, .lyrics == .c5, etc.

I'm not quite sure how to approach fixing this.  I'd like to make minimal changes to the HTML file, if possible, so that the CSS file can still be used to customize presentation.

**Comment by mpogue2** [comment]
By the way, HTML tidy is available as C source code and designed to be used as in library form, if we do need to go down that path.  I was looking at tidy as a possible way to clean up bad HTML on the input side.

**Comment by mpogue2** [comment]
And, one other thing I learned in playing around is that the SqView cuesheets use <L> and </L> for lyrics, in addition to using <FONT class="lyrics"> and </FONT>.  That's not legal HTML, so I had to strip it out (using sed) before I could run tidy.

This worked for me:
```
sed 's/<l>//g' summer.html | sed 's=<\l>==g' | tidy -i -w 160 -u -b -c > summer2.html
```

**Comment by danlyke** [comment]
I wish we could get meta information about the contents of the buffer out in some other way than just `toHtml()`. We might need to parse the HTML in-memory and rewrite it. I'd like to reduce it to something super simple and semantic.

Does it look like the CSS is copied in from the cuesheet verbatim? If we could just load the CSS files and do a `htmlString.replace(cssFileContents,referenceToCssFile)`, that might help.

**Comment by danlyke** [comment]
Oh, ugh, no: The style is per tag, not in a block at the top. The editor is throwing away all the semantic meaning.

So we can use reading the document into the text editor as our "HTML Tidy", but coming back out we need to figure out how to do something with it. Especially since we're going to need to be able to automagically replace choreography in it on the fly.

**Comment by danlyke** [comment]
Oh, ugh, no: The style is per tag, not in a block at the top. The editor is throwing away all the semantic meaning.

So we can use reading the document into the text editor as our "HTML Tidy", but coming back out we need to figure out how to do something with it. Especially since we're going to need to be able to automagically replace choreography in it on the fly.

Fixed the issue with the cuesheet not loading with 302e5cfa7aa417b42f800bda9fc2ff536743f346 and merged the master branch back in.

**Comment by danlyke** [comment]
Lyrics->Revert Edits added in 572efa931db874488fcac97437f9618aa86a37a8

**Comment by danlyke** [comment]
Thinking about saveability... I think we may need to track cuesheets in the database, and add a checkbox or toggle that says "always overwrite this file with the current changes". Still thinking on this...

**Comment by mpogue2** [comment]
> Yes, semantic meaning is lost.  But all is not lost! I've been thinking (always a dangerous thing :-)...I do think that because many of the semantic tags are non overlapping that I can infer the semantics from the presentation styles at the back end, especially with a little careful processing on the front end going in.  Yes it involves using tidy to help and non-greedy regexes.  I will take a whack at this over the long weekend and see if I can solve it.

I am ok with some transformations that simplify and regularize the HTML.  But I do want to retain the semantic info.

**Comment by mpogue2** [comment]
> It's more complex than that.  We already tell it which CSS to use, but then it translates it on the way in to eliminate the semantics leaving just inline CSS.
I think I can get around this...stay tuned!

**Comment by mpogue2** [comment]
Dan, editable lyrics branch is not compilable right now.  Did you forget to commit changes to main window.h and mainwindow.ui?

**Comment by mpogue2** [comment]
Yep, I think that's the problem...so, I have created a branch one commit back from that one, where I am now playing with HTML-TIDY.  

It is so cool!  I haven't gotten to the hard part yet (preserving semantic markup), but I am getting more comfortable with the tidy API (setup to affect Mac OS X only so far).  The output HTML is so much easier to read.

Example of tidied-up HTML (same input file as above):
```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
  <HEAD>
    <META content="text/html; charset=utf-8" http-equiv="content-type">
    <TITLE>
      Summer Sounds
    </TITLE>
    <LINK href="cuesheet2.css" rel="STYLESHEET" type="text/css">
  </HEAD>
  <BODY>
    <P class="hdr">
      (16 beat intro)
    </P>
    <P class="hdr">
      OPENER (Basic)
    </P><SPAN style="font-weight: bold;">Circle left!</SPAN><BR>
    <SPAN class="lyrics">Listen to the music of the carousel<BR>
    The ting-a-ling-a-ling of the ice cream bell</SPAN><BR>
    Do an Allemande Left your Corner, Turn a Right hand 'round your own<BR>
    And the Men Star Left and you roll it, roll it around you run<BR>
    Turn the Partner Right, then you go Left Allemande<BR>
    Swing a little girl and you Promenade the land<BR>
    <SPAN class="lyrics">Here...come...summer sounds, the summer sounds I love</SPAN><FONT class="lyrics"><BR></FONT><BR>
    <P class="hdr">
      FIGURE 1 (Basic, corner progression)
    </P>(HEADS) Star Thru, Pass Thru<BR>
```

**Comment by mpogue2** [comment]
Check out 662fa2aad86dc9da46447e64fdcd000ebe2e9acb .
I am successfully recovering the semantics for hdr and lyrics.  More tomorrow, but I think the methodology is sound.  On input, I tidy.  On output, I tidy again, get rid of cruft, and regex replace the presentation stuff with semantic stuff.

Currently assumes that semantics do not overlap, e.g. can't be both header and title at the same time.  The editor will need to enforce this (somehow?).  I'll worry about that problem later.  Worst case, you'll clear everything and re enter all the semantics.  PITA, but it will do until I can fix this.

**Comment by danlyke** [comment]
Whoops. Fixed.

**Comment by mpogue2** [comment]
OK, I think I have the semantic recovery stuff working, as of 94dd10ae08ca1ce28e4600f906b24a52c64ab56d .  

As long as the colors are distinct in the /lyrics/cuesheet2.css global style file, QTextEdit will retain them, and the original semantics can be recovered.  I have tested round-tripping several times on Mac OS X, with no loss of semantic information.

Dependency on html-tidy, which canonicalizes the tags, making it possible to use QRegExp to modify the HTML.  Might already be on your system? (or you might have to build it).  For MAC OS X, like libbass, I built it in a subdirectory (checked into the repository), and copied over the DLLs to the test123 directory (checked in as binary), where they are built into the .app file at build time.  

Couple of things I did not get to:
- Allow just having cursor on a line to toggle the entire block when button is pressed (like MRichTextEditor)
- Make Header/Lyrics/etc work like B/I, in that the states of the checkable buttons match the current style where the cursor is
- a style remover button (I think you already wrote this in the other branch?)
- make Title, Artist, and Label singletons, and update the < TITLE > when Title button is pressed

- shortcuts Cmd-B and Cmd-I.  Note that Cmd-I is already in use.
- Lyrics button doesn't go down when pressed.  It's not a native control, but maybe I have it set up wrong.
- Save As...still takes me to the wrong spot.
- Right click context menus for cut/copy/paste/bold/italic/header/lyrics/etc.
- Lyrics menu options:  submenu?
- editing ID3 tag, if HTML
- making the subdirectory Make file so that it auto-builds the html-tidy stuff as part of the normal build process
- using the USER'S colors for editing (right now, it uses the /lyrics/cuesheet2.css colors, and they are hard-coded into the app)
- separating stuff out into logically separate files (we're at 7000 lines in the mainwindow.cpp file)
- documentation
- testing, testing, testing
- Lasting peace in the Middle East

OK, I think this demonstrates that the semantic stuff can be recovered...wanna try it on Linux, and see if you like it?  Or maybe you can see a better way to do this?  I'm agnostic as to how we solve this, just excited that it appears to be solvable.

**Comment by mpogue2** [comment]
Here's what the prototype looks like on Mac OS X:
![screenshot 2017-06-30 19 59 37](https://user-images.githubusercontent.com/6316003/27758655-d2c51c04-5dce-11e7-97a0-f2b49775cf06.png)

**Comment by danlyke** [comment]
Merged mpogue3 into editablelyrics

So on Linux, we can install the tidy libraries on the system. It seems like putting the include files for those in the project is the wrong direction, but I'm not sure how to manage external libraries on systems which don't have good package management (*cough* Mac Windows *cough*).

**Comment by mpogue2** [comment]
Did it work OK for you on Linux?

I am unsure how to deal with the external package problem.  If it's an off-the-shelf library, Linux can just use the pre-built package (Mac and Win might be able to, as well, via MacPorts or Brew on Mac, and who-know-what on Windows).  But, if we (for example) modify html-tidy to make it specific to SquareDesk, then how should this be handled?  As a subdir of our main project (but then, how to tie in the CMake process?)?  Or, do we just call it a dependency, treat it as a separate project that is installed (maybe with a different name?)?

Maybe libbass could be moved to a separate project, too (it currently copies in binaries for Win/Mac, in part because only binaries are available)?  

I'm open to your advice here...I don't know what the best solution is.

**Comment by danlyke** [comment]
On Sat, 01 Jul 2017 16:06:01 -0700
Mike Pogue <notifications@github.com> wrote:
> Did it work OK for you on Linux?

Yep. I had to remove some of your ifdefs.

> I am unsure how to deal with the external package problem.  If it's
> an off-the-shelf library, Linux can just use the pre-built package
> (Mac and Win might be able to, as well, via MacPorts or Brew on Mac,
> and who-know-what on Windows).  But, if we (for example) modify
> html-tidy to make it specific to SquareDesk, then how should this be
> handled?

Can we build using git submodules? If we do end up making changes, I
feel like we should be pushing those changes back upstream to the
original maintainers.

> Maybe libbass could be moved to a separate project, too (it currently
> copies in binaries for Win/Mac, in part because only binaries are
> available)?  

My current challenges with packaging on Debian revolve around figuring
out how to do libbass internally, I need to sit down and make libbass a
package...

**Comment by mpogue2** [comment]
I don't know what git submodules are...I'll take a look!  Yes, we should push changes back upstream, if we believe that the changes are of general interest.  SD changes, IMHO, are probably not in good enough shape to push upstream, and they are of limited use outside SDP.

**Comment by mpogue2** [comment]
A little help around lyrics editing:  013be0d034c4f1f558e181808fdaa489bbfca6a7 .  Feel free to edit this to make it better.

**Comment by danlyke** [comment]
I expect that once we integrate SquareDesk directly rather than trying
to shell out to it, our changes will be more interesting.

**Comment by danlyke** [comment]
dcfb5875f51ba7417287f198f0e22c7686614bcb should solve the "Allowust having cursor on a line to toggle the entire block when button is pressed (like MRichTextEditor)"

I'm not seeing the styles change immediately on a ceder.net downloaded/derived cuesheet, I assume that something's not loading .../cuesheet2.css .

On "Save As...still takes me to the wrong spot.": Two possible issues:

1. Now that I'm saving and restoring the cuesheet save path properly (I think, on Linux at least, though I'm still only saving it on a proper close), is there a possibility that you have a bad path there?

2. If not, can you print out the path we're passing to the file save dialog in:

```
    QString filename = QFileDialog::getSaveFileName(this,
                                                    tr("Select Cue Sheet"),
                                                    lastCuesheetSavePath + "/" + fi.completeBaseName() + ".html",
                                                    tr("HTML (*.html)"));
```
and see what it's suggesting?

On "using the USER'S colors for editing (right now, it uses the /lyrics/cuesheet2.css colors, and they are hard-coded into the app)": Does this mean we need a tab to define the contents  /lyrics/cuesheet2.css in the preferences dialog? Colors and font sizes for the different semantic blocks?

Aaaand: I'm still pondering save behavior, but am now becoming more sure that "Save As..." as a button in the toolbar is not the right way to go.

**Comment by danlyke** [comment]
Okay, the Linux problem with the lyrics/cuesheet2.css is, of course, that I didn't have one. Maybe detecting the lack of it and warning the user somehow?

**Comment by danlyke** [comment]
And I've got an HTML Tidy crash that I'm trying to figure out how to solve...

**Comment by mpogue2** [comment]
re: toggle the entire block -- cool!
re: lack of cuesheet2.css -- I just added it to the build (Mac OS X only, until it's time for Win and Linux).  On Mac OS X, it will go inside the app itself, so it's available for use, if one isn't found in the lyrics directory.  My current thinking is that we should use the user's .css file if we can, and the built-in .css file if we can't.  BUT, if we use the internal one, we should probably also use that when there isn't one OR we could copy it into the user's lyrics directory if there isn't one anywhere.  But, I'm not sure about this -- what do you think?
re: I'll take a look at why Save As isn't defaulting to the last-used directory yet.  Even if I had a bad string in there now, when I save it should overwrite it with a correct one, right?  We already do this for opening MP3 files and for playlists...I'll take a look at why this is different...
re: Preferences rather than cuesheet.css -- I'm not sure on this.  The entire SquareView world uses cuesheet2.css, so it would be nice to continue that.  But, it is clunky, and hard for people to use.  OTOH maybe people don't really change it.  I've seen many callers use SqView, and not one of them had changed cuesheet2.css ....so....I guess for now let's leave it there, and save the work associated with making it a preference.  In fact, as per above, we might want to stick on in, if the user doesn't have one.
re: HTML Tidy crash -- check your version number.  I think I pulled the latest one, and I wasn't thinking about the fact that you might already have a released one on Linux!

**Comment by mpogue2** [comment]
re: getting rid of the Save As... button -- Let's talk about this!  I no longer really see a reason to have a File > Open for MP3.  I don't think any of us uses it.  Should we just ditch the MP3 load/save and replace it with *Lyrics* Open/Save/Save As, using the standard shortcuts?  Or, should this go into the Lyrics menu?  I'm also wondering where to put the styling stuff (Bold, Italic, Header, Title, etc.) -- under the Edit menu, or under the Lyrics menu?

**Comment by mpogue2** [comment]
310239ee4dc69c85bea1ef6cb4225eb59b53855c fixes the persistence problem of the save directory for lyrics.  This must be a Linux-specific thing, because canonicalPath returns "." on Mac.  I replaced that with an explicit extraction of the directoryName, and it seems to work on Mac OS X now.  Let me know if it works the same way on Linux!

**Comment by mpogue2** [comment]
OK, now I am using this thing for real work, and I'm hitting some problems:
- I am composing a new singing call, but none of the normal key sequences (e.g. Space to Play) work, because I'm editing inside the text box.  
- And, I don't have the Play and Stop buttons anymore.  So, I am stuck with Cmd-P and Cmd-., which are not very intuitive...I think I'd rather press buttons.
- ESC doesn't work to get me out of the text box.
- I really need a way to Clear Formatting.  Revert edits didn't actually revert the edits (from the lyrics template, because my singing call doesn't have lyrics yet), and if it did, it would wipe out everything, not just the few areas I wanted to clear.
- Cmd-LeftArrow doesn't work.  It's eaten up as a shortcut for Skip Ahead/Back before it gets to the text box (where it is needed for editing).
- First time I loaded the new singing call (which doesn't have lyrics) I got the template (correct).  But, the template wasn't colored like it should be (incorrect).  I think it was missing the cuesheet2.css being loaded.  After I loaded it (via a DIFFERENT song that had a cuesheet2.css), then the new song has colors now.  I think we should probably just load the internal cuesheet2.css (I thought I had done that, but I guess not!)  I'm actually OK with not using the user's cuesheet for editing right now...
- I want to listen to the same section over and over while I am composing lyrics.  I think user-defined (e.g. draggable) loops would probably give me that.  Maybe there's some key combination that could mean Go Back To Where I Last Started?

**Comment by mpogue2** [comment]
- When I Save As.., it reloads the cuesheet, which brings it back to the top (when I was editing at the bottom)....

**Comment by mpogue2** [comment]
Yes yes yes to your suggestion about Save.  The Normal SAVE command (Cmd-S) should work (and probably Cmd-Shift-S, too).  I am used to saving my work often via Cmd-S, and the reload-goes-to-the-top plus lack of Cmd-S makes it take way more time than it should....

Eatin' our own dogfood here. :-)

**Comment by danlyke** [comment]
Yeah, I think an explicit "now I'm editing, now I'm not" check box
would be good.

I'm also a big believer that the whole notion of "save" is a broken
artifact of making people doing their own cache management, but that
probably means implicit save points and versioning... And then we've
rewritten git.

**Comment by danlyke** [comment]
I realize that we've done a bunch of work here, but...

What if we just used http://doc.qt.io/qt-5/qfilesystemwatcher.html to watch for the cue sheet change and beef up the .txt file markup/markdown capabilities? Seems like alt-tabbing over to Emacs might actually be a better editing process for me... If we want to keep what we've done so far, make it so that it's an option to pop up that editor in a new window (or as a separate app), with normal "File->Save" sort of capabilities?

**Comment by mpogue2** [comment]
re: explicit Editing vs Not -- I agree...maybe a lock that when clicked is unlocked?

re: reinventing git...I think most people understand Saving nowadays.  Apple tried to get rid of it, and every app that uses that "you can't save, only duplicate" paradigm is crazy difficult to use, because it breaks the Save paradigm that isn't ideal but everybody knows it.

re: Markup/Markdown -- I don't think that Markup/Markdown is better than WYSIWYG for 99% of users (not including programmers like us, of course!).  And, I think we would still need the lingua franca of HTML in the SqView/VicCeder style (since everybody has already standardized on it).  I think most users would like our SDP cue sheets to be usable by SqView, if possible...So, that would then take us down a path of having a translator that roundtrips between Markup and SQVHTML, not a trivial thing either.  So, I think that would argue for Markup being an "alternate" editor, rather than the "primary" editor, and the lingua franca is still SQVHTML in both cases.

If you would like to (for example) create a new Preference, something like 
"Use Editor for cuesheets: {WYSIWYG INTERNAL|X|Y|Z}  [Choose...]", I think that would be totally great for advanced users!  But, I think we'd still need the WYSIWYG editor for 99% of users -- so it doesn't save much work.  I'm not worried about the amount of work already done on the WYSIWYG version (or the amount of additional work required to make it work well).  My gut feel is that 99% will want to use WYSIWYG (rather than Markdown or raw HTML), and if the other 1% prefer an HTML editor, or VI or emacs, then providing a way for those advanced users to get to it (e.g. with a Preference) is totally OK.

I don't think it would be viable to have everybody switch to using Markdown *in place of* SQVHTML.  

Another option might be to allow ".md" files *in addition* to ".html" files.  That looks to be the standard for Markdown.... but Markdown doesn't have verbs for Lyrics, Title, Author, so it would necessarily be non-standard Markdown.   But, it would be editable using a simpler text editor!

One other thing I thought of is having the ability to switch between plain text (.htm as text view) and rich text (WYSIWYG view) editor.  An .md editor might want to do the same.

I dunno -- did those thoughts help move the discussion forward, or did that just make everything murky? :-)

**Comment by mpogue2** [comment]
Also, one more thought:  It might be very nice to have a file watcher that updates the songTable when a new song is added to or deleted from or renamed in the music directory from outside the program.  Or, a new cuesheet added/deleted/modified anywhere from outside the program.   A couple of times I bought music, stuck it in, and realized that I needed to restart the app to get it visible.  Not a huge thing, just a minor annoyance.  

So, whatever way we go on lyric editing, I vote yes on the file watcher! :-)

**Comment by danlyke** [comment]
WARNING: c5d9e8bbaef2602daf83879b3af226b33fc89056 will break the Mac and Windows use of the tidy library. I think it should be relatively easy to fix, and is a step in the right direction, but I wanted that at the top...

This adds a checkbox to let you turn on and off editing of cuesheets, and one for auto-saving. This is known wrong and broken, but/and it's an additional something to play with.

**Comment by mpogue2** [comment]
I'll work on that tonight!  I am motivated to do an Alpha release this weekend that includes the Lyrics Editor capability, so I might be poking around on the UX this weekend...

**Comment by mpogue2** [comment]
html-tidy fixed by caf305265d27ea4081245b2936cfb9924ab9421a.
Added Lyrics > Print Lyrics... and Lyrics > Save/Save As... in c03e1e45ec48bc4538d68d21f6242f899a9ce647.  I left the buttons there for now, but I think we might want to convert them to QToolButtons, if they're gonna stay around.

I think I'm gonna close this one for now.  Future enhancements can be opened as new feature requests.  The basics are there now, and I think they are usable.  User-testing will tell us whether that's really true or not!
